src/mesh/duplicate_nodes.h

```diff
@@ -0,0 +1,151 @@
+ * E.g.:
+ * 
+ * Consider a domain containing fracture like this:
+ * +-----+-----+
+ * |     |     |
+ * |     +     |
+ * |           |
+ * |           |
+ * |           |
+ * +-----------+
+ * 
+ * with the mesh nodes numbered as follows:
+ * 0-----1-----2
+ * |     |     |
+ * |     3     |
+ * |           |
+ * |           |
+ * |           |
+ * 4-----------5
+ * 
+ * The class duplicates node 1 and node 3 because in these nodes FE functions can be discontinuous
+ * (schematic view):
+ * 0-6   1   7-2
+ * |  \  |  /  |
+ * |   \ 3 /   |
+ * |    \ /    |
+ * |     8     |
+ * |           |
+ * 4-----------5
+ * 
+ * Now, nodes 1,6,7 have the same coordinates but belong to different groups of elements
+ * and analogously nodes 3 and 8.
+ * 
+ * 
+ * The structure is used in DOF handler to distribute dofs for FE spaces sharing dofs
+ * between elements.
+ * 
+ * TODO: Currently we do not create faces of faces, i.e. 1d edges of tetrahedra are not
+ * available. This should be implemented if we need to distribute dofs on edges.
+ * 
+ */
+class DuplicateNodes {
+public:
+  
+  DuplicateNodes(Mesh *mesh);
+
+  // Getters (see private section for explanation).  
+  Mesh *mesh() const { return mesh_; }
+  
+  const std::vector<unsigned int> &nodes() const { return nodes_; }
+  const std::vector<unsigned int> &node_dim() const { return node_dim_; }
+  
+  const std::vector<MeshObject<0> > &points() const { return points_; }
+  const std::vector<MeshObject<1> > &lines() const { return lines_; }
+  const std::vector<MeshObject<2> > &triangles() const { return triangles_; }
+  const std::vector<MeshObject<3> > &tetras() const { return tetras_; }
+  
+  const std::vector<unsigned int> &obj_4_el() const { return obj_4_el_; }
+  const std::vector<unsigned int> &obj_4_edg() const { return obj_4_edg_; }
+  const std::vector<unsigned int> &obj_4_node() const { return obj_4_node_; }
+  
+  
+private:
+  
+  /// Initialize the vector of nodes from mesh.
+  void init_nodes();
+  
+  /// Initialize objects from mesh edges.
+  void init_from_edges();
+  
+  /// Initialize objects from mesh elements.
+  void init_from_elements();
+  
+  /// Duplicate nodes that are lying on interfaces with fractures.
+  void duplicate_nodes();
+  
+  
+  /// The mesh object.
+  Mesh *mesh_;
+  
+  /// Internal indices of nodes (including duplicated nodes).
+  std::vector<unsigned int> nodes_;
+  
+  /// Vector of space dimensions of elements using the particular node.
+  std::vector<unsigned int> node_dim_;
+  
+  /// Internal indices of points (0d elements).
+  std::vector<MeshObject<0> > points_;
+  
+  /// Internal indices of lines (1d elements).
+  std::vector<MeshObject<1> > lines_;
+  
+  /// Internal indices of triangles (2d elements).
+  std::vector<MeshObject<2> > triangles_;
+  
+  /// Internal indices of tetrahedra (3d elements).
+  std::vector<MeshObject<3> > tetras_;
+  
+  /// Vector of object indices for each mesh element.
+  std::vector<unsigned int> obj_4_el_;
```

Comment:
No. Elements are all together, objects are grouped by dimension. So obj_4_el_[i] returns the index of i-th element in the vector lines_, triangles_ or tetras_, depending on the element dimension.